plugins/backstage-plugin-coder/src/api/CoderClient.ts

```diff
@@ -0,0 +1,392 @@
+  private readonly authApi: CoderAuthApi;
+
+  private readonly options: CoderClientConfigOptions;
+  private readonly snapshotManager: StateSnapshotManager<CoderClientSnapshot>;
+  private readonly axiosInterceptorId: number;
+  private readonly abortController: AbortController;
+
+  private latestProxyEndpoint: string;
+  readonly sdkApi: BackstageCoderSdkApi;
+
+  /* ***************************************************************************
+   * There is some funky (but necessary) stuff going on in this class - a lot of
+   * the class methods are passed directly to other systems. Just to be on the
+   * safe side, all methods (public and private) should be defined as arrow
+   * functions, to ensure the methods can't ever lose their `this` contexts
+   *
+   * This technically defeats some of the memory optimizations you would
+   * normally get with class methods (arrow methods will be rebuilt from
+   * scratch each time the class is instantiated), but because CoderClient will
+   * likely be instantiated only once for the entire app's lifecycle, that won't
+   * matter much at all
+   ****************************************************************************/
+
+  constructor(inputs: CoderClientConstructorInputs) {
+    const { apis, ...options } = inputs;
+    const { discoveryApi, identityApi, authApi } = apis;
+
+    // The "easy setup" part - initialize internal properties
+    this.identityApi = identityApi;
+    this.discoveryApi = discoveryApi;
+    this.authApi = authApi;
+    this.latestProxyEndpoint = '';
+    this.options = { ...defaultCoderClientConfigOptions, ...options };
+
+    /**
+     * Wire up SDK API namespace.
+     *
+     * @todo All methods are defined locally in the class, but this should
+     * eventually be updated so that 99% of methods come from the SDK, with a
+     * few extra methods patched in for Backstage convenience
+     */
+    this.sdkApi = {
+      getUserLoginType: this.getUserLoginType,
+      getWorkspaceBuildParameters: this.getWorkspaceBuildParameters,
+      getWorkspaces: this.getWorkspaces,
+      getWorkspacesByRepo: this.getWorkspacesByRepo,
+    };
+
+    // Wire up Backstage APIs and Axios to be aware of each other
+    this.abortController = new AbortController();
+    this.axiosInterceptorId = axiosInstance.interceptors.request.use(
+      this.interceptAxiosRequest,
+    );
+
+    // Hook up snapshot manager so that external systems can be made aware when
+    // state changes, all in a render-safe way
+    this.snapshotManager = new StateSnapshotManager({
+      initialSnapshot: this.prepareNewStateSnapshot(),
+    });
+
+    // Set up logic for syncing client snapshots to auth state changes
+    this.authApi.subscribe(newAuthSnapshot => {
+      const latestClientSnapshot = this.getStateSnapshot();
+      if (newAuthSnapshot.isTokenValid !== latestClientSnapshot.isAuthValid) {
+        this.notifySubscriptionsOfStateChange();
+      }
+    });
```

Comment:
This is the main thing I was talking about in the `StateSnapshotManager` PR with how it's a little clunky to get two snapshots to be related to each other